scalding-core/src/main/scala/com/twitter/scalding/typed/WritePartitioner.scala

```diff
@@ -0,0 +1,297 @@
+package com.twitter.scalding.typed
+
+import com.twitter.scalding.Execution
+import com.stripe.dagon.{ Dag, Id, Rule, Memoize, FunctionK }
+import org.slf4j.LoggerFactory
+import scala.language.higherKinds
+
+object WritePartitioner {
+  private[this] val logger = LoggerFactory.getLogger(getClass)
+
+  type PairK[F[_], G[_], T] = (F[T], G[T])
+
+  /**
+   * This breaks a job at all the places it explicitly fans out,
+   * (and currently after each reduce/join).
```

Comment:
Do we have enough optimization rules to pull `.map` functions into the reducer ? or does it make more sense to explicitly not do that and let user code do mapValues/etc.. to stay in the reducer?

Reply:
This partitioning is not yet optimal. It may wind up in more steps that cascading would have taken. It would be nice to get it so that is never the case, and we could just turn it on all the time.

Since the optimizer works, for now, on the same nodes as TypedPipe, we can't push map into a reduce because it can change the keys. We can push mapValues/flatMapValues/filterKeys. We are actually not *yet* doing mapValues and flatMapValues. I just didn't get around to it. Since adding rules is not binary incompatible, I've been punting on adding a few of the last rules.